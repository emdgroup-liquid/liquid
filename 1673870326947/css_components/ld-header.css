.ld-header,:host(.ld-header){--ld-header-height:3.125rem;--ld-header-max-width:90rem;--ld-header-col:var(--ld-col-wht);--ld-header-bg-col:var(--ld-thm-primary);--ld-header-box-shadow:var(--ld-shadow-stacked);background-color:var(--ld-header-bg-col);box-shadow:var(--ld-header-box-shadow);color:var(--ld-header-col);display:flex;justify-content:center;transition:transform .1s ease-in-out;width:100%}.ld-header.ld-header--sticky,:host(.ld-header.ld-header--sticky){position:sticky;top:0;z-index:1}.ld-header.ld-header--hidden,:host(.ld-header.ld-header--hidden){transform:translateY(-100%)}.ld-header__container{align-items:center;box-sizing:border-box;display:flex;height:var(--ld-header-height);max-width:var(--ld-header-max-width);padding-left:var(--ld-sp-16);width:100%}.ld-header__container>:not([mode=ghost]):not(.ld-button--ghost):not(.ld-header__logo):not(.ld-header__logo-wrapper),::slotted(:not([mode=ghost]):not(.ld-button--ghost):not(.ld-header__logo):not(.ld-header__logo-wrapper)){margin-right:var(--ld-sp-16)}.ld-header__container>.ld-button--ghost,.ld-header__container>[mode=ghost],::slotted(.ld-button--ghost),::slotted([mode=ghost]){margin:0 var(--ld-sp-12) 0 calc(var(--ld-sp-4)*-1)}.ld-header__logo-wrapper{color:inherit;display:flex}.ld-header__logo-wrapper ::slotted(*){margin:0}.ld-header__logo{--ld-icon-size-md:2.4rem;color:var(--ld-thm-warning);display:block;margin:-.2rem;margin-right:calc(var(--ld-sp-16) - .2rem)}.ld-header_site-name{white-space:nowrap}.ld-header__grow{flex:1 0 auto}:host .ld-header__grow{margin:0}