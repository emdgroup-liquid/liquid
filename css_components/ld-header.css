.ld-header,:host(.ld-header){--ld-header-height:3.125rem;--ld-header-max-width:90rem;background-color:var(--ld-thm-primary);box-shadow:var(--ld-shadow-stacked);color:var(--ld-col-wht);display:flex;justify-content:center;width:100%;transition:transform .1s ease-in-out}.ld-header.ld-header--sticky,:host(.ld-header.ld-header--sticky){position:sticky;top:0;z-index:1}.ld-header.ld-header--hidden,:host(.ld-header.ld-header--hidden){transform:translateY(calc(var(--ld-header-height)*-1))}.ld-header__container{align-items:center;box-sizing:border-box;display:flex;height:var(--ld-header-height);max-width:var(--ld-header-max-width);padding-left:var(--ld-sp-16);width:100%}.ld-header__container>:not([mode=ghost]):not(.ld-button--ghost):not(.ld-header__logo):not(.ld-header__logo-wrapper),::slotted(:not([mode=ghost]):not(.ld-button--ghost):not(.ld-header__logo):not(.ld-header__logo-wrapper)){margin-right:var(--ld-sp-16)}.ld-header__container>.ld-button--ghost,::slotted(.ld-button--ghost),.ld-header__container>[mode=ghost],::slotted([mode=ghost]){margin:0 var(--ld-sp-8) 0 calc(var(--ld-sp-8)*-1)}.ld-header__logo-wrapper{color:inherit;display:flex;}.ld-header__logo-wrapper ::slotted(*){margin:0}.ld-header__logo{--ld-icon-size-md:2.4rem;color:var(--ld-thm-warning);display:block;margin:-.2rem;margin-right:calc(var(--ld-sp-16) - .2rem)}.ld-header_site-name{white-space:nowrap}.ld-header__grow{flex:1 0 auto;}:host .ld-header__grow{margin:0}