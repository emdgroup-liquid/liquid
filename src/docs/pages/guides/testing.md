---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

