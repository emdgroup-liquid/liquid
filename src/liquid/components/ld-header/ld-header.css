:host {
  --ld-header-height: 3.125rem;

  background-color: var(--ld-thm-primary);
  color: var(--ld-col-wht);
  display: flex;
  width: 100%;

  header {
    align-items: center;
    display: flex;
    height: var(--ld-header-height);
    justify-content: space-between;
    margin-right: var(--ld-sp-16);
    max-width: 80rem;
    width: 100%;
  }
}

::slotted(ld-button:not(:first-child):not([mode='ghost'])),
::slotted(ld-menu) {
  margin-left: var(--ld-sp-16);
}

::slotted(ld-button[mode='ghost']) {
  margin: 0 calc(var(--ld-sp-8) * -1) 0 var(--ld-sp-8);
}

.ld-header__logo {
  color: inherit;
  margin-left: var(--ld-sp-16);

  ld-icon {
    --ld-icon-size-md: 2rem;
    color: var(--ld-thm-warning);
  }
}

.ld-header__grow-wrapper {
  align-items: center;
  display: flex;
  flex: 1;

  > :where(*) {
    margin-left: var(--ld-sp-16);
  }
}

.ld-header__start,
.ld-header__end {
  display: flex;
}
