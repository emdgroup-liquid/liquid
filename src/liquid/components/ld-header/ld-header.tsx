import { Component, Host, h, Prop, Element } from '@stencil/core'
import { getClassNames } from 'src/liquid/utils/getClassNames'

/**
 * @slot end - Items on the right side of the header.
 * @slot logo - Custom logo.
 * @slot menu - Main menu.
 * @slot start - Items on the left side of the header.
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part container - Actual header element that limits the width of the header content
 * @part logo - `ld-icon` element containing the default logo
 * @part site-name - `ld-typo` element containing the site name
 */
@Component({
  assetsDirs: ['assets'],
  tag: 'ld-header',
  styleUrl: 'ld-header.css',
  shadow: true,
})
export class LdHeader {
  @Element() el: HTMLElement

  /** Hide the header when the user scrolls down and show it again, when the user scrolls up. */
  @Prop() hideOnScroll = false

  /** Title attribute of the logo link. */
  @Prop() logoTitle?: string

  /** URL that the logo links to. */
  @Prop() logoUrl?: string

  /** Make the header sticky. */
  @Prop() sticky = false

  /** Name shown on the right side of the logo. */
  @Prop() siteName?: string

  componentDidLoad() {
    const ldMenu = this.el.querySelector('ld-menu')

    if (ldMenu) {
      ldMenu.orientation = 'horizontal'
    }

    this.el
      .querySelectorAll<HTMLLdButtonElement>('ld-header > ld-button')
      .forEach((ldButton) => {
        ldButton.size = 'sm'
        ldButton.brandColor = true

        if (ldButton.mode === 'ghost') {
          const textInButton = ldButton.textContent.trim()
          const iconInButton = ldButton.querySelector<HTMLElement>('.ld-icon')
          const innerButton = ldButton.shadowRoot.querySelector('button')

          if (iconInButton && !textInButton) {
            iconInButton.style.setProperty('--ld-icon-size-lg', '1.5rem')
            iconInButton.style.setProperty('--ld-icon-size-md', '1.5rem')
            iconInButton.style.setProperty('--ld-icon-size-sm', '1.5rem')
            innerButton.style.setProperty('--ld-button-padding-x-sm', '0')
            innerButton.style.setProperty('--ld-button-padding-y-sm', '0')
          }
        }
      })

    this.el.querySelectorAll('ld-header > .ld-button').forEach((cssButton) => {
      cssButton.classList.add('ld-button--brand-color')
    })
  }

  render() {
    const cl = getClassNames([
      'ld-header',
      this.hideOnScroll && 'ld-header--hide-on-scroll',
      this.sticky && 'ld-header--sticky',
    ])

    return (
      <Host class={cl} role="banner">
        <header part="container">
          <slot name="start" />
          {this.logoUrl ? (
            <a
              class="ld-header__logo"
              href={this.logoUrl}
              title={this.logoTitle}
            >
              <slot name="logo">
                <ld-icon name="m" part="logo" />
              </slot>
            </a>
          ) : (
            <div class="ld-header__logo">
              <slot name="logo">
                <ld-icon name="m" part="logo" title={this.logoTitle} />
              </slot>
            </div>
          )}
          <div class="ld-header__grow-wrapper">
            {this.siteName && (
              <ld-typo part="site-name" tag="div" variant="h5">
                {this.siteName}
              </ld-typo>
            )}
            <slot name="menu" />
          </div>
          <slot name="end" />
        </header>
      </Host>
    )
  }
}
