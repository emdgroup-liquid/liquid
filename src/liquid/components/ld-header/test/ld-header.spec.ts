jest.mock("../../../utils/cloneAttributes");
import { newSpecPage } from "@stencil/core/testing";
import { LdHeader } from "../ld-header";

describe("ld-header", () => {
  it("renders", async () => {
    const page = await newSpecPage({
      components: [LdHeader],
      html: `<ld-header></ld-header>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("renders with site name", async () => {
    const page = await newSpecPage({
      components: [LdHeader],
      html: `
      <ld-header site-name="Liquid Oxygen"></ld-header>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("renders with linked logo", async () => {
    const page = await newSpecPage({
      components: [LdHeader],
      html: `
      <ld-header logo-title="Home" logo-url="#"></ld-header>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("renders with logo title", async () => {
    const page = await newSpecPage({
      components: [LdHeader],
      html: `
      <ld-header logo-title="Logo"></ld-header>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("renders with content", async () => {
    const page = await newSpecPage({
      components: [LdHeader],
      html: `
      <ld-header>
        <ld-typo>Liquid Oxygen</ld-typo>
      </ld-header>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("renders sticky", async () => {
    const page = await newSpecPage({
      components: [LdHeader],
      html: `<ld-header sticky></ld-header>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("renders hidden", async () => {
    const page = await newSpecPage({
      components: [LdHeader],
      html: `<ld-header sticky hidden></ld-header>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("renders hide-on-scroll", async () => {
    const page = await newSpecPage({
      components: [LdHeader],
      html: `<ld-header sticky hide-on-scroll></ld-header>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("does not throw when disconnecting before hydration", () => {
    const component = new LdHeader();
    component.disconnectedCallback();
  });
});
