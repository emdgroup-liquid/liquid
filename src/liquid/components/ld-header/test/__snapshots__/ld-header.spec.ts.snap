// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-header renders 1`] = `
<ld-header class="ld-header" logo-title="Home" role="banner" site-name="Liquid Oxygen">
  <mock:shadow-root>
    <header part="container">
      <slot name="start"></slot>
      <div class="ld-header__logo">
        <slot name="logo">
          <ld-icon name="m" part="logo" title="Home"></ld-icon>
        </slot>
      </div>
      <div class="ld-header__grow-wrapper">
        <ld-typo part="site-name" tag="div" variant="h5">
          Liquid Oxygen
        </ld-typo>
        <slot name="menu"></slot>
      </div>
      <slot name="end"></slot>
    </header>
  </mock:shadow-root>
</ld-header>
`;

exports[`ld-header renders with burger menu button 1`] = `
<ld-header class="ld-header" role="banner" site-name="Liquid Oxygen">
  <mock:shadow-root>
    <header part="container">
      <slot name="start"></slot>
      <div class="ld-header__logo">
        <slot name="logo">
          <ld-icon class="ld-icon" part="logo" role="presentation">
            <mock:shadow-root>
              Not Found
            </mock:shadow-root>
          </ld-icon>
        </slot>
      </div>
      <div class="ld-header__grow-wrapper">
        <ld-typo part="site-name" tag="div" variant="h5">
          Liquid Oxygen
        </ld-typo>
        <slot name="menu"></slot>
      </div>
      <slot name="end"></slot>
    </header>
  </mock:shadow-root>
  <ld-button slot="start" type="button">
    <mock:shadow-root>
      <button aria-live="polite" class="ld-button" part="button focusable">
        <slot></slot>
      </button>
    </mock:shadow-root>
    <ld-icon class="ld-icon" name="burger-menu" role="presentation">
      <mock:shadow-root>
        Not Found
      </mock:shadow-root>
    </ld-icon>
  </ld-button>
</ld-header>
`;

exports[`ld-header renders with buttons 1`] = `
<ld-header class="ld-header" role="banner" site-name="Liquid Oxygen">
  <mock:shadow-root>
    <header part="container">
      <slot name="start"></slot>
      <div class="ld-header__logo">
        <slot name="logo">
          <ld-icon name="m" part="logo"></ld-icon>
        </slot>
      </div>
      <div class="ld-header__grow-wrapper">
        <ld-typo part="site-name" tag="div" variant="h5">
          Liquid Oxygen
        </ld-typo>
        <slot name="menu"></slot>
      </div>
      <slot name="end"></slot>
    </header>
  </mock:shadow-root>
  <ld-button slot="end" type="button">
    <mock:shadow-root>
      <button aria-live="polite" class="ld-button" part="button focusable">
        <slot></slot>
      </button>
    </mock:shadow-root>
    Register
  </ld-button>
  <ld-button mode="secondary" slot="end" type="button">
    <mock:shadow-root>
      <button aria-live="polite" class="ld-button ld-button--secondary" part="button focusable">
        <slot></slot>
      </button>
    </mock:shadow-root>
    Login
  </ld-button>
</ld-header>
`;

exports[`ld-header renders with clickable logo 1`] = `
<ld-header class="ld-header" logo-title="Home" logo-url="#" role="banner" site-name="Liquid Oxygen">
  <mock:shadow-root>
    <header part="container">
      <slot name="start"></slot>
      <a class="ld-header__logo" href="#" title="Home">
        <slot name="logo">
          <ld-icon name="m" part="logo"></ld-icon>
        </slot>
      </a>
      <div class="ld-header__grow-wrapper">
        <ld-typo part="site-name" tag="div" variant="h5">
          Liquid Oxygen
        </ld-typo>
        <slot name="menu"></slot>
      </div>
      <slot name="end"></slot>
    </header>
  </mock:shadow-root>
</ld-header>
`;

exports[`ld-header renders with custom clickable logo 1`] = `
<ld-header class="ld-header" logo-title="Home" logo-url="#" role="banner" site-name="Liquid Oxygen">
  <mock:shadow-root>
    <header part="container">
      <slot name="start"></slot>
      <a class="ld-header__logo" href="#" title="Home">
        <slot name="logo">
          <ld-icon class="ld-icon" part="logo" role="presentation">
            <mock:shadow-root>
              Not Found
            </mock:shadow-root>
          </ld-icon>
        </slot>
      </a>
      <div class="ld-header__grow-wrapper">
        <ld-typo part="site-name" tag="div" variant="h5">
          Liquid Oxygen
        </ld-typo>
        <slot name="menu"></slot>
      </div>
      <slot name="end"></slot>
    </header>
  </mock:shadow-root>
  <ld-icon class="ld-icon ld-icon--lg" name="rocket" role="presentation" size="lg" slot="logo">
    <mock:shadow-root>
      Not Found
    </mock:shadow-root>
  </ld-icon>
</ld-header>
`;

exports[`ld-header renders with custom logo 1`] = `
<ld-header class="ld-header" logo-title="Home" role="banner" site-name="Liquid Oxygen">
  <mock:shadow-root>
    <header part="container">
      <slot name="start"></slot>
      <div class="ld-header__logo">
        <slot name="logo">
          <ld-icon class="ld-icon" part="logo" role="presentation" title="Home">
            <mock:shadow-root>
              Not Found
            </mock:shadow-root>
          </ld-icon>
        </slot>
      </div>
      <div class="ld-header__grow-wrapper">
        <ld-typo part="site-name" tag="div" variant="h5">
          Liquid Oxygen
        </ld-typo>
        <slot name="menu"></slot>
      </div>
      <slot name="end"></slot>
    </header>
  </mock:shadow-root>
  <ld-icon class="ld-icon ld-icon--lg" name="rocket" role="presentation" size="lg" slot="logo">
    <mock:shadow-root>
      Not Found
    </mock:shadow-root>
  </ld-icon>
</ld-header>
`;

exports[`ld-header renders with menu 1`] = `
<ld-header class="ld-header" role="banner" site-name="Liquid Oxygen">
  <mock:shadow-root>
    <header part="container">
      <slot name="start"></slot>
      <div class="ld-header__logo">
        <slot name="logo">
          <ld-icon name="m" part="logo"></ld-icon>
        </slot>
      </div>
      <div class="ld-header__grow-wrapper">
        <ld-typo part="site-name" tag="div" variant="h5">
          Liquid Oxygen
        </ld-typo>
        <slot name="menu"></slot>
      </div>
      <slot name="end"></slot>
    </header>
  </mock:shadow-root>
  <ld-menu class="ld-menu ld-menu--vertical" role="navigation" slot="menu">
    <mock:shadow-root>
      <ul class="ld-menu__items" part="list">
        <slot></slot>
      </ul>
    </mock:shadow-root>
    <ld-menu-item href="#">
      <mock:shadow-root>
        <li>
          <a class="ld-menu-item ld-menu-item--list" href="#" part="link">
            <slot></slot>
          </a>
        </li>
      </mock:shadow-root>
      Menu Item 1
      <ld-menu-item href="#">
        <mock:shadow-root>
          <li>
            <a class="ld-menu-item ld-menu-item--list" href="#" part="link">
              <slot></slot>
            </a>
          </li>
        </mock:shadow-root>
        Menu Item 2
        <ld-menu-item href="#">
          <mock:shadow-root>
            <li>
              <a class="ld-menu-item ld-menu-item--list" href="#" part="link">
                <slot></slot>
              </a>
            </li>
          </mock:shadow-root>
          Menu Item 3
        </ld-menu-item>
      </ld-menu-item>
    </ld-menu-item>
  </ld-menu>
</ld-header>
`;

exports[`ld-header renders without site name 1`] = `
<ld-header class="ld-header" role="banner">
  <mock:shadow-root>
    <header part="container">
      <slot name="start"></slot>
      <div class="ld-header__logo">
        <slot name="logo">
          <ld-icon name="m" part="logo"></ld-icon>
        </slot>
      </div>
      <div class="ld-header__grow-wrapper">
        <slot name="menu"></slot>
      </div>
      <slot name="end"></slot>
    </header>
  </mock:shadow-root>
</ld-header>
`;
